From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jianfeng Liu <Jianfeng.Liu@cixtech.com>
Date: Mon, 20 Jul 2026 21:06:30 +0800
Subject: DPTSW-10531-1: USB: CDNSP: fix bug for incorrect cdns version

For CDNSP_CONTROLLER_V2 controller, we should write and read the
correct simulate register.

Signed-off-by: Jianfeng Liu <Jianfeng.Liu@cixtech.com>
---
 drivers/usb/cdns3/core.c | 3 +++
 drivers/usb/cdns3/drd.c  | 6 +++++-
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/cdns3/core.c b/drivers/usb/cdns3/core.c
index 111111111111..222222222222 100644
--- a/drivers/usb/cdns3/core.c
+++ b/drivers/usb/cdns3/core.c
@@ -405,6 +405,9 @@ static int cdns_role_set(struct usb_role_switch *sw, enum usb_role role)
 	if (role != USB_ROLE_NONE)
 		cdnsp_platform_reset(cdns->dev);
 
+	if (cdns->version == CDNSP_CONTROLLER_V2)
+		writel(1, &cdns->otg_cdnsp_regs->simulate);
+
 	ret = cdns_role_start(cdns, role);
 	if (ret)
 		dev_err(cdns->dev, "set role %d has failed\n", role);
diff --git a/drivers/usb/cdns3/drd.c b/drivers/usb/cdns3/drd.c
index 111111111111..222222222222 100644
--- a/drivers/usb/cdns3/drd.c
+++ b/drivers/usb/cdns3/drd.c
@@ -429,6 +429,7 @@ int cdns_drd_init(struct cdns *cdns)
 			cdns->otg_irq_regs = (struct cdns_otg_irq_regs __iomem *)
 					      &cdns->otg_cdnsp_regs->ien;
 			cdns->version  = CDNSP_CONTROLLER_V2;
+			writel(1, &cdns->otg_cdnsp_regs->simulate);
 		} else if (OTG_CDNS3_CHECK_DID(state)) {
 			cdns->otg_irq_regs = (struct cdns_otg_irq_regs __iomem *)
 					      &cdns->otg_v1_regs->ien;
@@ -503,9 +504,12 @@ bool cdns_power_is_lost(struct cdns *cdns)
 	if (cdns->version == CDNS3_CONTROLLER_V0) {
 		if (!(readl(&cdns->otg_v0_regs->simulate) & BIT(0)))
 			return true;
-	} else {
+	} else if (cdns->version == CDNS3_CONTROLLER_V1) {
 		if (!(readl(&cdns->otg_v1_regs->simulate) & BIT(0)))
 			return true;
+	} else if (cdns->version == CDNSP_CONTROLLER_V2) {
+		if (!(readl(&cdns->otg_cdnsp_regs->simulate) & BIT(0)))
+			return true;
 	}
 	return false;
 }
-- 
Cixtech

